### PR TITLE
Make it possible to enable CORS

### LINKED ISSUE #177: CORS is not enabled for BinderHub API
I'd like to use BinderHub's API to start kernels from another webpage.

When trying to start a notebook server using the BinderHub API, the browser will error because the Access-Control-Allow-Origin HTTP header is missing.

screenshot 2017-10-05 19 48 03

I believe this should be a simple implementation of [`set_default_headers`](https://stackoverflow.com/a/35259440) in the `base.py:BaseHandler` class but I'm still trying to get binderhub set up locally so I can't test it myself.

In addition, the current release of `notebook` (5.1.0) is missing a header that allows starting kernels from another domain. I fixed it in https://github.com/jupyter/notebook/pull/2876 but is there something we can do to get the fix in for the current deployment of Binder?

Happy to talk more about this (in person works too).

Related to #13.

### PR DESCRIPTION
**Pull request by minrk**
enables access to the eventsource endpoint from other websites, not just scripts.

closes #177

Potential questions:

- Do we want to enable this by default (as this PR does)? Or should it be more locked down by default and opt-in on the mybinder.org deployment? (**answer: not enabled by default**)

cc @SamLau95

CORS is not enabled by default. To enable CORS from any origin in `values.yaml`:

```yaml
cors: &cors
  allowOrigin: '*'

jupyterhub:
  extraConfigMap:
    cors: *cors
```

### CODE REVIEW

**Comment by yuvipanda** [comment]
@minrk yeah, I think the default should have this off and we should enable this for mybinder.org

Otherwise lgtm!

**Comment by minrk** [comment]
Makes sense. I will do that in the morning:

1. switch default to off
2. expose it in helm chart
3. update mybinder.org-deploy to enable it

**Comment by minrk** [comment]
It would be ideal if we can make the origin setting the same for notebook servers and binderhub itself. I'll see if that's doable, and do the simple thing first, if not.

**Comment by minrk** [comment]
Okay, this should be set. The simplest way to enable CORS:

```yaml
cors: &cors
  allowOrigin: '*'

jupyterhub:
  extraConfigMap:
    cors: *cors
```

I was looking up info about yaml anchors to try to avoid setting this in more than one place and I couldn't find it. In a given `values.yaml`, anchors can be used to avoid setting *values* twice, but the anchor link must be established in each `values.yaml` (i.e. user values.yaml doesn't inherit anchors from the chart's values.yaml). The only way I think we could get this to be truly a single location is to put it in `Values.global`, but that would require teaching the hub chart about something that belongs in the Hub configmap coming from Values.global.

**Comment by SamLau95** [comment]
Thanks @minrk ! I won't be able to immediately make API calls to beta.mybinder.org after this PR is merged; I'll have to wait until you guys update the deployment, is that right?

**Comment by minrk** [comment]
@SamLau95 yes, but that is usually very soon after the PR lands with our current workflow. I'll let you know when it's deployed on beta.

File: helm-chart/binderhub/values.yaml [hunk 1]
```diff
@@ -35,6 +35,18 @@ github:
   # ...or access_token
   accessToken:
 
+# have to set cors.allowOrigin twice:
+# once in top-level cors.allowOrigin,
+# and again in jupyterhub.hub.extraConfigMap.cors.allowOrigin
+
+# &cors anchor allows us to remove redundancy in this values.yaml,
```
Comment (willingc) [inline helm-chart/binderhub/values.yaml]: Perhaps:

```
# Using YAML anchors, `&cors` for the first appearance and `*cors` for subsequent
# appearances, allows us to remove redundancy in this (BinderHub) `values.yaml`.
# The anchors do not extend beyond this file.
# As such, users must set `cors` separately for their own user `values.yaml` for notebooks. 
# `cors` will be set separately in the user `values.yaml` and binderhub`values.yaml`.
# The same anchor pattern (`&cors`, `*cors`) can be used in the user `values.yaml`.
```

Comment (willingc) [reply helm-chart/binderhub/values.yaml]: @minrk A bit more context above on YAML anchors will make the pattern more understandable.

Comment (minrk) [reply helm-chart/binderhub/values.yaml]: Thanks! added your comment.

**Comment by willingc** [comment]
Thanks @minrk. Nice addition about anchors not extending beyond the file.

**Pull request merged**